File: HPWorkpath_v1.6.3/Samples/ExampleAPIServices/source/SuppliesSample/src/com/hp/workpath/sample/suppliessample/task/SuppliesTask.java
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.suppliessample.task;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;

import com.hp.workpath.api.Result;
import com.hp.workpath.api.supplies.SuppliesService;
import com.hp.workpath.api.supplies.supplyinfo.Supply;
import com.hp.workpath.sample.suppliessample.Logger;
import com.hp.workpath.sample.suppliessample.fragment.ResponseInterface;

import java.lang.ref.WeakReference;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SuppliesTask {

    private WeakReference<Context> mContextRef;
    private ResponseInterface mResponseInterface;
    private Result mResult;

    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Handler handler = new Handler(Looper.getMainLooper());

    public SuppliesTask(Context context, ResponseInterface responseInterface) {
        this.mContextRef = new WeakReference<>(context);
        this.mResponseInterface = responseInterface;
        this.mResult = new Result();
    }

    public void taskExecute() {
        try {
            executor.execute(new Runnable() {
                @Override
                public void run() {
                    try {
                        onPostExecute(SuppliesService.getSuppliesInfo(mContextRef.get(), mResult));
                    } catch (Throwable t) {
                        Logger.showResult(null, "SuppliesService.getSuppliesInfo" + t.getMessage());
                        executor.shutdown();
                        onPostExecute(null);
                    }

                }
            });
        } catch (Exception e) {
            Logger.showResult(null, "SuppliesService.getSuppliesInfo" + e.getMessage());
            onPostExecute(null);
            executor.shutdown();
        }
    }

    private void onPostExecute(List<Supply> supplyList) {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (mResult.getCode() == Result.RESULT_OK && supplyList != null) {
                    mResponseInterface.success(supplyList);
                } else {
                    mResponseInterface.failure("SuppliesService.getSuppliesInfo", mResult);
                }
            }
        });
    }
}

File: HPWorkpath_v1.6.3/Samples_Kotlin/ExampleAPIServices/source/SuppliesSample/src/com/hp/workpath/sample/suppliessample/task/SuppliesTask.kt
// Copyright 2018 HP Development Company, L.P.
// SPDX-License-Identifier: MIT
package com.hp.workpath.sample.suppliessample.task

import android.content.Context
import com.hp.workpath.api.Result
import com.hp.workpath.api.supplies.SuppliesService
import com.hp.workpath.api.supplies.supplyinfo.Supply
import com.hp.workpath.sample.suppliessample.fragment.ResponseInterface
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.lang.ref.WeakReference

class SuppliesTask(context: Context, responseInterface: ResponseInterface) {
    private val mContextRef: WeakReference<Context> = WeakReference(context)
    private val mResponseInterface: ResponseInterface = responseInterface
    private var mThrowable: Throwable? = null
    private var mResult: Result = Result()

    suspend fun execute() {
        var supplies: List<Supply>? = null
        mContextRef.get()?.run {
            try {
                supplies = SuppliesService.getSuppliesInfo(this, mResult)
            } catch (t: Throwable) {
                mThrowable = t
            }
        }
        onPostExecute(supplies)
    }

    private suspend fun onPostExecute(supplyList: List<Supply>?) {
        withContext(Dispatchers.Main) {
            if (mResult.code == Result.RESULT_OK && supplyList != null) {
                mResponseInterface.success(supplyList)
            } else {
                mThrowable?.run {
                    mResponseInterface.failed("SuppliesService.getSuppliesInfo $message", null)
                } ?: run {
                    mResponseInterface.failed("SuppliesService.getSuppliesInfo", mResult)
                }
            }
        }
    }
}